src/pam/oslogin_sshca.cc

```diff
@@ -0,0 +1,322 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <oslogin_sshca.h>
+#include <openbsd.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct sshca_type {
+  const char *type;
+  int (*skip_custom_fields)(char **buff, size_t *blen);
+} sshca_type;
+
+static int sshca_dsa_skip_fields(char **buff, size_t *blen);
+static int sshca_ecdsa_skip_fields(char **buff, size_t *blen);
+static int sshca_ed25519_skip_fields(char **buff, size_t *blen);
+static int sshca_rsa_skip_fields(char **buff, size_t *blen);
+
+static sshca_type sshca_impl[] = {
+    {"ecdsa-sha2-nistp256-cert-v01@openssh.com", sshca_ecdsa_skip_fields},
+    {"ecdsa-sha2-nistp384-cert-v01@openssh.com", sshca_ecdsa_skip_fields},
+    {"ecdsa-sha2-nistp521-cert-v01@openssh.com", sshca_ecdsa_skip_fields},
+    {"rsa-sha2-256-cert-v01@openssh.com", sshca_rsa_skip_fields},
+    {"rsa-sha2-512-cert-v01@openssh.com", sshca_rsa_skip_fields},
+    {"ssh-dss-cert-v01@openssh.com", sshca_dsa_skip_fields},
+    {"ssh-ed25519-cert-v01@openssh.com", sshca_ed25519_skip_fields},
+    {"ssh-rsa-cert-v01@openssh.com", sshca_rsa_skip_fields},
+    { },
+};
+
+static int
+sshca_get_string(char **buff, size_t *blen, char **ptr, size_t *len_ptr) {
```

Comment:
Nit: I notice that all over this file, you use the return code to indicate success. Our style guide recommends "Prefer using return values over output parameters: they improve readability, and often provide the same or better performance." cf. https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

But I understand that in plain C, it is hard to use a string as a return value (returning `char *` requires the receiver to figure out the length of the string all over again), so feel free to ignore this nit. ;)